### DISCUSSION TITLE
Gestalt 0.25.2 supports MapLiteralDecoder.java

### DISCUSSION THREAD

**Discussion by credmond-git** [openTCS/opentcs#138] [General]
If you upgrade to [Gestalt v0.25.2](https://github.com/gestalt-config/gestalt/releases/tag/v0.25.2) it includes your [MapLiteralDecoder.java](https://github.com/openTCS/opentcs/blob/master/openTCS-Impl-Configuration-gestalt/src/main/java/org/opentcs/configuration/gestalt/decoders/MapLiteralDecoder.java) as part of the existing map decoder. 
Feel free to test and use the Gestalt implementation to reduce the amount of custom code you need to support. 

What is the goal of the [ClassPathDecoder.java](https://github.com/openTCS/opentcs/blob/master/openTCS-Impl-Configuration-gestalt/src/main/java/org/opentcs/configuration/gestalt/decoders/ClassPathDecoder.java)? How is it used and what does it produce? I would consider taking PR or use the code if it doesn't overlap with other decoders.

**Comment by LennysLounge** [comment]
> If you upgrade to [Gestalt v0.25.2](https://github.com/gestalt-config/gestalt/releases/tag/v0.25.2) it includes your [MapLiteralDecoder.java](https://github.com/openTCS/opentcs/blob/master/openTCS-Impl-Configuration-gestalt/src/main/java/org/opentcs/configuration/gestalt/decoders/MapLiteralDecoder.java) as part of the existing map decoder.
Feel free to test and use the Gestalt implementation to reduce the amount of custom code you need to support.

 That is very nice. Although i wouldnt mind supporting this ourselves its nice that we dont have to. And im sure your code is a little more robust than mine is. Im goning to try to update to the newest version if i can.
 
 

> What is the goal of the [ClassPathDecoder.java](https://github.com/openTCS/opentcs/blob/master/openTCS-Impl-Configuration-gestalt/src/main/java/org/opentcs/configuration/gestalt/decoders/ClassPathDecoder.java)? How is it used and what does it produce?

In openTCS it is possible to create custom location and vehicle themes by implementing the theme interfaces. If a theme implementation is on the class path it is then possible to choose which theme to use by specifying the fully qualified class name in the configuration.
The output of this deocer is then a `Class<? extends LocationTheme>` or `Class<? extends VehicleTheme>` class object.
The configuration interface is here: [PlantOverviewOperatingApplicationConfiguration](https://github.com/openTCS/opentcs/blob/07006bb772308e597207e0c4724e0e06fd8320a7/openTCS-OperationsDesk/src/main/java/org/opentcs/operationsdesk/util/PlantOverviewOperatingApplicationConfiguration.java#L78)
and a example configuration is here: [opentcs-operationsdesk-defaults-baseline.properties](https://github.com/openTCS/opentcs/blob/07006bb772308e597207e0c4724e0e06fd8320a7/openTCS-OperationsDesk/src/main/resources/org/opentcs/operationsdesk/distribution/config/opentcs-operationsdesk-defaults-baseline.properties#L16)

I did not find a decoder that could do this and it was honestly easy enough to implement myself.

**Reply by credmond-git** [reply]
Thanks for the detailed description. The Map functionality was something that fit really well into the library, so i thought it would be a great addition. 

The ClassPathDecoder is interesting and useful functionality, but it seems like it is more of a DI framework providing a class than a configuration. I am happy that Gestalt can suit your needs for this, but i am not sure i will integrate this functionality into Gestalt for now. I am glad it was easy to add the functionality.
